Shading Elasticsearch

slovit opened the pull request.

Fgerthoffert commented:
@junjun-zhang @btiernay @shanewilson @andricDu any comments on this PR ?

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+## Introduction
+
+'Shading' or package renaming(relocation) is a process of creating an [uber-jar](http://imagej.net/Uber-JAR) which contains its dependencies and package names of some of the dependendencies are renamed.
```
Space after "renaming"

btiernay replied:
"a.k.a class relocation". Would have "class relocation" point to https://maven.apache.org/components/plugins/maven-shade-plugin/examples/class-relocation.html

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+'Shading' or package renaming(relocation) is a process of creating an [uber-jar](http://imagej.net/Uber-JAR) which contains its dependencies and package names of some of the dependendencies are renamed.
+
+In this blog post I will provide instructions how to create an Elasticsearch jar file with 'shaded' dependencies. This operation is often required when you use some libraries the Elasticsearch depends on, but version of the library differs. In such a case your project might not compile or run. A solution to the problem is to create a custom Elasticsearch distribution jar file which has the proble ...
```
"that Elasticsearch"

btiernay replied:
" but version of the library differs" -> " but the version of the library conflicts in breaking ways"

btiernay replied:
Try to add a space before all of your opening `(`

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+## Maven project configuration
+
+We will use the Apache Maven to create the jar files, so let's prepare the standard maven directories layout.
```
"We will use the Apache Maven" -> "We will use Apache Maven"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+## Building
+
+To build 'shaded' jar files and install them in the local maven repository use the following command in the project root directory:
```
"the shaded"

btiernay replied:
> the project root directory

Would indicate what it would be in our example.

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+There is a problem related to testing with Elasticsearch. Sometimes it is preferable to be able to query against a running instance of Elasticsearch in unit tests, because mocking it too cumbersome or even not possible. There are cases (like development of own [domain-specific language](https://github.com/icgc-dcc/dcc-portal/tree/develop/dcc-portal-pql)), when it is not suitable to query against  ...
+
+The drawback of using the `ESIntegTestCase` is that it requires additional configuration when you would like to run scripts. Modules which support particular query language, for example `Painless` or `Groovy` should be added separately. Those modules are not available in the Central Maven repository, however they are provided with the Elasticsearch artifact.
```
"Elasticsearch artifact" -> "Elasticsearch distribution"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+#### elasticsearch-test-shaded
+
+There is a problem related to testing with Elasticsearch. Sometimes it is preferable to be able to query against a running instance of Elasticsearch in unit tests, because mocking it too cumbersome or even not possible. There are cases (like development of own [domain-specific language](https://github.com/icgc-dcc/dcc-portal/tree/develop/dcc-portal-pql)), when it is not suitable to query against  ...
```
"not recomemnded" -> spelling

btiernay replied:
"of own" -> "of your own"

btiernay replied:
"use provided" -> "use the provided"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+#### elasticsearch-test-shaded
+
+There is a problem related to testing with Elasticsearch. Sometimes it is preferable to be able to query against a running instance of Elasticsearch in unit tests, because mocking it too cumbersome or even not possible. There are cases (like development of own [domain-specific language](https://github.com/icgc-dcc/dcc-portal/tree/develop/dcc-portal-pql)), when it is not suitable to query against  ...
```
"3-rd" -> 3rd or third

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+mvn clean deploy
+~~~
+
+Most likely the remote repository requires authentication. The following lines should be added to the `~/.m2/settings.xml`:
```
"added to the" -> "added to" / "added to your"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+To deploy the artifacts to a remote repository the use
+
+~~~
```
"```shell"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+</dependencies>
+~~~
+
+Next, we need to download ES distribution and extract `lang-painless` jar file with all it's dependencies and include those to the classpath. This is accomplished with `com.googlecode.maven-download-plugin` and `com.googlecode.addjars-maven-plugin`. Plugins configuration is pretty self-describing.
```
"extract `lang-painless`" -> "extract the `lang-painless`"

btiernay replied:
"it's dependencies" -> "its dependencies"

btiernay replied:
"Plugins configuration" -> "The plugin's configuration"

btiernay replied:
"self-describing." -> "self-describing:"

btiernay commented:
This is a very useful post and well written! I recommend posting a link on the ES discuss after publication.

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+Next, we will define shade plugin configuration. 
+
+We include only Elasticsearch packages and the relocated dependencies in shaded jar files. All the other Elasticsearch dependencies are promoted as transitive ones. Such a setup provides a better dependencies control, as it's possible to exclude some of the dependencies or override their version.
```
"override their version." -> "override their version downstream in your project if the need arise"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+#### Parent pom
+
+We start with a parent pom and put there all the common information we need as well as a `maven-shade-plugin` configuration. It easy to maintain plugin configuration if it is stored in one location.
```
"and put there all the common" -> "to centralize the common"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+author: vitalii_slobodianyk
+date: 2016-11-25
+categories: vitalii_slobodianyk
+tags:
```
`- Jar Hell` :)

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+'Shading' or package renaming(relocation) is a process of creating an [uber-jar](http://imagej.net/Uber-JAR) which contains its dependencies and package names of some of the dependendencies are renamed.
+
+In this blog post I will provide instructions how to create an Elasticsearch jar file with 'shaded' dependencies. This operation is often required when you use some libraries the Elasticsearch depends on, but version of the library differs. In such a case your project might not compile or run. A solution to the problem is to create a custom Elasticsearch distribution jar file which has the proble ...
```
Would introduce the concept of "Jar Hell" here, perhaps linking to something like  https://en.wikipedia.org/wiki/Java_Classloader#JAR_hell

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+
+We start with a parent pom and put there all the common information we need as well as a `maven-shade-plugin` configuration. It easy to maintain plugin configuration if it is stored in one location.
+
+Define some common variables and settings we'll use:
```
"Define" -> " To define"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+</properties>
+~~~
+
+Next, we will define shade plugin configuration. 
```
"define shade" -> "define the `maven-shade-plugin`

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+~~~
+
+
+It is important to note that `joda` packages should not be shaded if you use scripts written in `Painless` language. It is a default Elasticsearch scripting language developed by the Elasticsearch team. It relies on Java data types and uses `joda`'s types for date. If you relocate it `Painless` script will fail to compile.
```
"It is a default" -> "It is the default"

btiernay replied:
"script will" -> "query scripts will"

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+~~~xml
+<artifactSet>
+  <includes>
+    <include>org.icgc.dcc:*</include>
```
Maybe make note of this?

btiernay commented on _posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md:
```diff
@@ -0,0 +1,350 @@
+~~~xml
+<distributionManagement>
+  <repository>
+    <id>dcc-dependencies</id>
```
Maybe generalize this, and talk about the general need to have place to host these artifacts.

slovit commented:
@btiernay @Fgerthoffert Addressed all the comments except the purpose of an own Artifactory. I could continue with that one, but that's another topic. Moreover, I'm working on ETL fixes now.
